Implement bilinear mixed modes traction separation law for cohesive zone model

jiangwen84 opened the pull request:
closes #18611

jiangwen84 committed:
implement bilinear mixed mode CZM
ref #18611

moosebuild commented:
Job [Documentation](https://civet.inl.gov/job/796779/) on 7a3009c wanted to post the following:

View the site [here](https://mooseframework.inl.gov/docs/PRs/18612/site)

This comment will be updated on new commits.

arovinelli commented on modules/tensor_mechanics/src/materials/cohesive_zone_model/BiLinearMixedMode.C:
```diff
@@ -0,0 +1,203 @@
+BiLinearMixedMode::validParams()
+{
+  InputParameters params = CZMComputeLocalTractionTotalBase::validParams();
+  params.addRequiredParam<Real>("stiffness", "Penalty stiffness.");
```
if this is the interpenetration penalty stiffness call the parameter and the variable `penalty_stiffness`

jiangwen84 replied:
done.

arovinelli reviewed (changes requested):
a few comments on the implementation of the material model. All the calcualtion are inside teh same method making it long and difficult to understand. Can you add methods such as: `computeDamage` for updating `_d`, `computeStiffnessAndDerivatives` for computing `_D` and `_dDdu`, this will make the code much easier to understand and maintain.

For what concerns test: I don't understand the reason why you unrolled all the tests in the `tests` file. It was organized to have one section testing the framework and other subsections to test material implementation. Your modification makes it much longer and more difficult to read. Unless there is another reason that I can't see, I would appreciate if you could simply add your new test in a new subsection.

Test, I only see normal test, can you add a sliding test. I would suggest adding it. Furthermore, I don't see any Jacobian check. If the jacobian is supposed to be correct please add a test similar to the `3DC_jacobian`, if the jacobian is approximated no need for the test but write it down in the doc.

jiangwen84 commented:
> a few comments on the implementation of the material model. All the calcualtion are inside teh same method making it long and difficult to understand. Can you add methods such as: `computeDamage` for updating `_d`, `computeStiffnessAndDerivatives` for computing `_D` and `_dDdu`, this will make the code much easier to understand and maintain.
> 
> For what concerns test: I don't understand the reason why you unrolled all the tests in the `tests` file. It was organized to have one section testing the framework and other subsections to test material implementation. Your modification makes it much longer and more difficult to read. Unless there is another reason that I can't see, I would appreciate if you could simply add your new test in a new subsection.
> 
> Test, I only see normal test, can you add a sliding test. I would suggest adding it. Furthermore, I don't see any Jacobian check. If the jacobian is supposed to be correct please add a test similar to the `3DC_jacobian`, if the jacobian is approximated no need for the test but write it down in the doc.

Regarding the `tests` file, I only add a new section for my staff. The other changes you saw is due to `hit` formatting. I do not think we should keep adding more `md` and `issues number` to the original `czm_framewrok` block. It is not clear which test corresponds to which issue and design. I prefer a new separate block.

Your existing `Jacobian` tests did not actually test if the hand-coded Jacobian is correct. The CSVDiff cannot guarantee you will have an exact Jacobian. You should at lease use `type = PetscJacobianTester`

jiangwen84 commented on modules/tensor_mechanics/include/materials/cohesive_zone_model/BiLinearMixedMode.h:
```diff
@@ -0,0 +1,89 @@
+  /// onset shear seperation
+  const Real _delta_shear0;
+
+  /// effective displacement at damage initiation
+  MaterialProperty<Real> & _eff_disp_damage_init;
+
+  /// effective displacement at full degradation
+  MaterialProperty<Real> & _eff_disp_full_degradation;
```
Those names are used in the cited paper, so I think it is fine. 

The reasons to make them material properties: 1. in case some users would like to lag them by using the old value. 2. We can output them to exodus file.

jiangwen84 commented on modules/tensor_mechanics/src/materials/cohesive_zone_model/BiLinearMixedMode.C:
```diff
@@ -0,0 +1,203 @@
+      use_old ? _maximum_mixed_mode_relative_displacement_old[_qp]
+              : mixed_mode_relative_displacement;
+
+  Real d = 0.0;
+
+  if (_maximum_mixed_mode_relative_displacement[_qp] *
+          (_eff_disp_full_degradation[_qp] - _eff_disp_damage_init[_qp]) >
+      0.0)
+    d = _eff_disp_full_degradation[_qp] *
+        (_maximum_mixed_mode_relative_displacement[_qp] - _eff_disp_damage_init[_qp]) /
+        (_maximum_mixed_mode_relative_displacement[_qp] *
+         (_eff_disp_full_degradation[_qp] - _eff_disp_damage_init[_qp]));
+
+  if (d < 0)
+    d = 0.0;
+  else if (d > 1)
+    d = 1.0;
+
+  d = (d + _viscosity * _d_old[_qp] / _dt) / (_viscosity / _dt + 1.0);
+
+  if (d > _d_old[_qp])
+    _d[_qp] = d;
+  else
+  {
+    _d[_qp] = _d_old[_qp];
+    use_old = true;
+  }
```
The function won't be very general and I do not see its value. 

But I add some more comments to guide users to understand the codes.

jiangwen84 commented on modules/tensor_mechanics/include/materials/cohesive_zone_model/BiLinearMixedMode.h:
```diff
@@ -0,0 +1,89 @@
+  /// viscous regularization
+  const Real _viscosity;
+
+  /// critical Mode I fracture toughness
+  const Real _GI_C;
+
+  /// critical Mode II fracture toughness
+  const Real _GII_C;
```
done

jiangwen84 commented on modules/tensor_mechanics/include/materials/cohesive_zone_model/BiLinearMixedMode.h:
```diff
@@ -0,0 +1,89 @@
+  /// penalty elastic stiffness
+  const Real _stiffness;
+
+  /// damage variable
+  MaterialProperty<Real> & _d;
+
+  /// old damage variable value
+  const MaterialProperty<Real> & _d_old;
```
done

jiangwen84 commented on modules/tensor_mechanics/src/materials/cohesive_zone_model/BiLinearMixedMode.C:
```diff
@@ -0,0 +1,203 @@
+
+    if (!use_old)
+    {
+      RealVectorValue dmax_du(
+          _interface_displacement_jump[_qp](0) > 0.0
+              ? _interface_displacement_jump[_qp](0) /
+                    _maximum_mixed_mode_relative_displacement[_qp]
+              : 0.0,
+          _interface_displacement_jump[_qp](1) / _maximum_mixed_mode_relative_displacement[_qp],
+          _interface_displacement_jump[_qp](2) / _maximum_mixed_mode_relative_displacement[_qp]);
```
I think one line is come compact :-)

jiangwen84 commented:
> > a few comments on the implementation of the material model. All the calcualtion are inside teh same method making it long and difficult to understand. Can you add methods such as: `computeDamage` for updating `_d`, `computeStiffnessAndDerivatives` for computing `_D` and `_dDdu`, this will make the code much easier to understand and maintain.
> > For what concerns test: I don't understand the reason why you unrolled all the tests in the `tests` file. It was organized to have one section testing the framework and other subsections to test material implementation. Your modification makes it much longer and more difficult to read. Unless there is another reason that I can't see, I would appreciate if you could simply add your new test in a new subsection.
> > Test, I only see normal test, can you add a sliding test. I would suggest adding it. Furthermore, I don't see any Jacobian check. If the jacobian is supposed to be correct please add a test similar to the `3DC_jacobian`, if the jacobian is approximated no need for the test but write it down in the doc.
> 
> Regarding the `tests` file, I only add a new section for my staff. The other changes you saw is due to `hit` formatting. I do not think we should keep adding more `md` and `issues number` to the original `czm_framewrok` block. It is not clear which test corresponds to which issue and design. I prefer a new separate block.
> 
> Your existing `Jacobian` tests did not actually test if the hand-coded Jacobian is correct. The CSVDiff cannot guarantee you will have an exact Jacobian. You should at lease use `type = PetscJacobianTester`

I changed my mind. I will add this material to the `material` section in `tests` file. It looks nicer.

jiangwen84 commented:
@arovinelli Thanks for your review. I addressed most of your comments. A shear mode test was added. I tried `PetscJacobianTester`, but it does not work well. It seems it only test for the first nonlinear iteration. Maybe I miss some important steps. Because of that, I did not add any test for Jacobian. For what I tested for, the number of nonlinear iterations with the hand-coded Jacobian is the same as that with FD. 

Please let me know if you have any more comments.

arovinelli commented:
> @arovinelli Thanks for your review. I addressed most of your comments. A shear mode test was added. I tried `PetscJacobianTester`, but it does not work well. It seems it only test for the first nonlinear iteration. Maybe I miss some important steps. Because of that, I did not add any test for Jacobian. For what I tested for, the number of nonlinear iterations with the hand-coded Jacobian is the same as that with FD.
> 
> Please let me know if you have any more comments.

I'm fine with checking against FD results, which is actually what `petsc` tester is doing.

Also, up to now all CZM material classes start with the `CZM` prefix. I think we should continue using it so they are easier to find in the documentation

jiangwen84 commented:
> > @arovinelli Thanks for your review. I addressed most of your comments. A shear mode test was added. I tried `PetscJacobianTester`, but it does not work well. It seems it only test for the first nonlinear iteration. Maybe I miss some important steps. Because of that, I did not add any test for Jacobian. For what I tested for, the number of nonlinear iterations with the hand-coded Jacobian is the same as that with FD.
> > Please let me know if you have any more comments.
> 
> I'm fine with checking against FD results, which is actually what `petsc` tester is doing.
> 
> Also, up to now, all CZM material classes start with the `CZM` prefix. I think we should continue using it so they are easier to find in the documentation.

The `PETSc` tester is comparing the entries of the hand-coded Jacobian and FD Jacobian. I could not make it work for my case. 

`SalehaniIrani3DCTraction` is an exception for the naming?

arovinelli commented on modules/tensor_mechanics/include/materials/cohesive_zone_model/BiLinearMixedMode.h:
```diff
@@ -0,0 +1,89 @@
+ * described in Mixed-Mode Decohesion Finite Elements for the Simulation of Delamination in
+ *Composite Materials, Pedro P. Camanho and Carlos G. Davila, NASA/TM-2002-211737
+ **/
+class BiLinearMixedMode : public CZMComputeLocalTractionTotalBase
```
Should we rename`BiLinearMixedMode` to `CZMBiLinearMixedMode` to be consisten with the rest of CZM materials?

arovinelli commented:
> > > @arovinelli Thanks for your review. I addressed most of your comments. A shear mode test was added. I tried `PetscJacobianTester`, but it does not work well. It seems it only test for the first nonlinear iteration. Maybe I miss some important steps. Because of that, I did not add any test for Jacobian. For what I tested for, the number of nonlinear iterations with the hand-coded Jacobian is the same as that with FD.
> > > Please let me know if you have any more comments.
> > 
> > 
> > I'm fine with checking against FD results, which is actually what `petsc` tester is doing.
> > Also, up to now, all CZM material classes start with the `CZM` prefix. I think we should continue using it so they are easier to find in the documentation.
> 
> The `PETSc` tester is comparing the entries of the hand-coded Jacobian and FD Jacobian. I could not make it work for my case.
> 
> `SalehaniIrani3DCTraction` is an exception for the naming?

I didn't think about that... but it has the name traction in it, so why not `BiLinearMixedModeTraction`

jiangwen84 commented:
> > > > @arovinelli Thanks for your review. I addressed most of your comments. A shear mode test was added. I tried `PetscJacobianTester`, but it does not work well. It seems it only test for the first nonlinear iteration. Maybe I miss some important steps. Because of that, I did not add any test for Jacobian. For what I tested for, the number of nonlinear iterations with the hand-coded Jacobian is the same as that with FD.
> > > > Please let me know if you have any more comments.
> > > 
> > > 
> > > I'm fine with checking against FD results, which is actually what `petsc` tester is doing.
> > > Also, up to now, all CZM material classes start with the `CZM` prefix. I think we should continue using it so they are easier to find in the documentation.
> > 
> > 
> > The `PETSc` tester is comparing the entries of the hand-coded Jacobian and FD Jacobian. I could not make it work for my case.
> > `SalehaniIrani3DCTraction` is an exception for the naming?
> 
> I didn't think about that... but it has the name traction in it, so why not `BiLinearMixedModeTraction`?
> Or as we are still in early-stage we could rename `SalehaniIrani3DCTraction` to `CZMSalehaniIrani3DC`, whatever works best for the users i guess

I guess `Traction` somehow implies CZM model. I am fine with `BiLinearMixedModeTraction`.

arovinelli commented:
> > > > > @arovinelli Thanks for your review. I addressed most of your comments. A shear mode test was added. I tried `PetscJacobianTester`, but it does not work well. It seems it only test for the first nonlinear iteration. Maybe I miss some important steps. Because of that, I did not add any test for Jacobian. For what I tested for, the number of nonlinear iterations with the hand-coded Jacobian is the same as that with FD.
> > > > > Please let me know if you have any more comments.
> > > > 
> > > > 
> > > > I'm fine with checking against FD results, which is actually what `petsc` tester is doing.
> > > > Also, up to now, all CZM material classes start with the `CZM` prefix. I think we should continue using it so they are easier to find in the documentation.
> > > 
> > > 
> > > The `PETSc` tester is comparing the entries of the hand-coded Jacobian and FD Jacobian. I could not make it work for my case.
> > > `SalehaniIrani3DCTraction` is an exception for the naming?
> > 
> > 
> > I didn't think about that... but it has the name traction in it, so why not `BiLinearMixedModeTraction`?
> > Or as we are still in early-stage we could rename `SalehaniIrani3DCTraction` to `CZMSalehaniIrani3DC`, whatever works best for the users i guess
> 
> I guess `Traction` somehow implies CZM model. I am fine with `BiLinearMixedModeTraction`.

I'm fine with it, it would be nice to keep this suffix consistent in the future :)

jiangwen84 commented:
@arovinelli do you have more comments? If not, could you approve this PR? Thanks.

arovinelli commented on modules/tensor_mechanics/doc/content/source/materials/cohesive_zone_model/BiLinearMixedModeTraction.md:
```diff
@@ -0,0 +1,96 @@
+\end{equation}
+
+\begin{equation}
+D_sr =
```
typo `D_{sr}`

jiangwen84 replied:
done.

arovinelli commented on modules/tensor_mechanics/doc/content/source/materials/cohesive_zone_model/BiLinearMixedModeTraction.md:
```diff
@@ -0,0 +1,96 @@
+
+Cohesive zone models exhibiting softening behavior and stiffness degradation often lead to convergence difficulties in an implicit solver. The traction-separation laws can be regularized using viscosity as given below,
+\begin{equation}
+\dot{d}=\frac{1}{\mu}(d-d_v)
```
what is `d_v`, also can you mention how you are integrating this equation? is implicit, explicit or you have close form for it?

jiangwen84 replied:
done.

arovinelli reviewed (changes requested):
a few minor comments

arovinelli reviewed (approved):
if all tests pass, this can be merged

dschwen commented on modules/tensor_mechanics/src/materials/cohesive_zone_model/BiLinearMixedModeTraction.C:
```diff
@@ -0,0 +1,212 @@
+
+BiLinearMixedModeTraction::BiLinearMixedModeTraction(const InputParameters & parameters)
+  : CZMComputeLocalTractionTotalBase(parameters),
+    _scale_factor(isCoupled("normal_strength_scale_factor")
```
why not just default `normal_strength_scale_factor` to 1?

jiangwen84 replied:
done.

dschwen commented on modules/tensor_mechanics/src/materials/cohesive_zone_model/BiLinearMixedModeTraction.C:
```diff
@@ -0,0 +1,212 @@
+        (_maximum_mixed_mode_relative_displacement[_qp] *
+         (_eff_disp_full_degradation[_qp] - _eff_disp_damage_init[_qp]));
+
+  if (d < 0)
+    d = 0.0;
+  else if (d > 1)
+    d = 1.0;
```
```suggestion
  d = MathUtils::clamp(d, 0.0, 1.0);
```

jiangwen84 committed:
Update modules/tensor_mechanics/src/materials/cohesive_zone_model/BiLinearMixedModeTraction.C

Co-authored-by: Daniel Schwen <daniel@schwen.de>

jiangwen84 committed:
address comments
